### PR TITLE
introduce new train method for sq

### PR DESCRIPTION
**Pull request by LHT129**

### CODE REVIEW

**Review by ShawnShawnYou — APPROVED** [review]

File: src/quantization/scalar_quantization_trainer.cpp [hunk 1]
```diff
@@ -74,11 +76,45 @@ ScalarQuantizationTrainer::classic_train(const float* data,
...
 
+void
+ScalarQuantizationTrainer::trunc_bound_train(const float* data,
+                                             uint64_t count,
+                                             float* upper_bound,
+                                             float* lower_bound) const {
+    auto ignore_count = static_cast<uint64_t>(static_cast<float>(count - 1) * 0.001);
+
+    for (uint64_t i = 0; i < dim_; ++i) {
+        upper_bound[i] = std::numeric_limits<float>::lowest();
```
Comment (inabao) [inline src/quantization/scalar_quantization_trainer.cpp]: Unnecessary initialization

File: src/quantization/scalar_quantization_trainer.cpp [hunk 2]
```diff
@@ -74,11 +76,45 @@ ScalarQuantizationTrainer::classic_train(const float* data,
...
+        lower_bound[i] = std::numeric_limits<float>::max();
+        std::priority_queue<float, std::vector<float>, std::greater<>> heap_max;
+        std::priority_queue<float, std::vector<float>, std::less<>> heap_min;
+        heap_max.emplace(upper_bound[i]);
+        heap_min.emplace(lower_bound[i]);
```
Comment (inabao) [inline src/quantization/scalar_quantization_trainer.cpp]: No need for initial values in advance

File: src/quantization/scalar_quantization_trainer.cpp [hunk 3]
```diff
@@ -74,11 +76,45 @@ ScalarQuantizationTrainer::classic_train(const float* data,
...
+        for (uint64_t j = 0; j < count; ++j) {
+            auto value = data[j * dim_ + i];
+            if (value > heap_max.top() || heap_max.size() < ignore_count) {
```
Comment (inabao) [inline src/quantization/scalar_quantization_trainer.cpp:92]: heap_max.size() < ignore_count || value > heap_max.top()

File: src/quantization/scalar_quantization_trainer.cpp [hunk 4]
```diff
@@ -74,11 +76,45 @@ ScalarQuantizationTrainer::classic_train(const float* data,
     }
 }
 
+void
+ScalarQuantizationTrainer::trunc_bound_train(const float* data,
+                                             uint64_t count,
+                                             float* upper_bound,
+                                             float* lower_bound) const {
+    auto ignore_count = static_cast<uint64_t>(static_cast<float>(count - 1) * 0.001);
```
Comment (inabao) [inline src/quantization/scalar_quantization_trainer.cpp:84]: what happened when `count = 0`?

File: src/quantization/scalar_quantization_trainer.cpp (hunk 3 above)
Comment (LHT129) [reply src/quantization/scalar_quantization_trainer.cpp:92]: why ? heap_max.size() < ignore_count is more stringent

File: src/quantization/scalar_quantization_trainer.cpp (hunk 4 above)
Comment (LHT129) [reply src/quantization/scalar_quantization_trainer.cpp:84]: never

**Review by inabao — APPROVED** [review]
LGTM

**Pull request merged**